cpp/src/arrow/compute/api_vector.h

```diff
@@ -115,13 +114,11 @@ class ARROW_EXPORT SortOptions : public FunctionOptions {
   /// Note: Both classes contain the exact same information.  However,
   /// sort_options should only be used in a "function options" context while Ordering
   /// is used more generally.
-  Ordering AsOrdering() && { return Ordering(std::move(sort_keys), null_placement); }
-  Ordering AsOrdering() const& { return Ordering(sort_keys, null_placement); }
+  Ordering AsOrdering() && { return Ordering(std::move(sort_keys)); }
+  Ordering AsOrdering() const& { return Ordering(sort_keys); }
 
   /// Column key(s) to order by and how to order by these sort keys.
   std::vector<SortKey> sort_keys;
-  /// Whether nulls and NaNs are placed at the start or at the end
-  NullPlacement null_placement;
```

Comment:
Removing this member is a breaking change, which should be explicitly called out in the PR description. Note that instead, we could keep this member (maybe deprecating it) as an optional default (see for example the same demotion applied to SetLookupOptions::skip_nulls in https://github.com/apache/arrow/pull/36739/files#diff-6bc7ecec6a4f7bcefc2511cde3bd809340ad0d94bb8f7cc5f4994063c798f2faR313 )

Reply:
When I wanted to set the deprecated field to a std::optional member, I found some conflicts while retaining the original construction, for example:

```
SortOptions(std::vector<SortKey> sort_keys = {}, NullPlacement null_placement = NullPlacement::AtEnd) {
}
```

If the user calls SortOptions, use the following method:
```
auto null_placement = {SortKey{NullPlacement::AtEnd}, SortKey{}};
SortOptions(sort_keys, NullPlacement::AtStart)
```

Should I use SortKey's NullPlacement or NullPlacement::AtStart in the constructor?

Therefore, here I added the interface modification annotation and did not use std::optional.